### PR TITLE
Migration of ReliableIrohaChainListener

### PR DESCRIPTION
**Pull request by dolgopolovwork**
`ReliableIrohaChainListener` goes here.

### CODE REVIEW

**Review by Mingela — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -39,4 +40,60 @@ chain-adapter:
     - rmq
   volumes:
     - ../your change adapter storage/:/deploy/chain-adapter
-  ```
\ No newline at end of file
+```
+  
+## How to use
+`com.d3.chainadapter.client.ReliableIrohaChainListener` is the class that is used as a client for the service. The class may be obtained via JitPack:
```
Comment (LiraLemur) [inline README.md]: I'd omit "that is"

File: README.md [hunk 2]
```diff
@@ -39,4 +40,60 @@ chain-adapter:
...
+
+```groovy
+compile "com.github.d3ledger.chain-adapter:chain-adapter-client:$chain_adapter_client_version"
+``` 
+Typical workflow looks like the following:
```
Comment (LiraLemur) [inline README.md]: "looks like this" maybe?

File: README.md [hunk 3]
```diff
@@ -39,4 +40,60 @@ chain-adapter:
...
+
+1) First, you must create an instance of `ReliableIrohaChainListener` object. 
+2) Then you have to call the `getBlockObservable()` function that returns `Observable<Pair<BlockOuterClass.Block, () -> Unit>>` wrapped by `Result`(see [github.com/kittinunf/Result](https://github.com/kittinunf/Result)). The returned object may be used to register multiple subscribers.
```
Comment (LiraLemur) [inline README.md]: wrapped in?

File: README.md [hunk 4]
```diff
@@ -39,4 +40,60 @@ chain-adapter:
...
+The first component of `Pair` is an Iroha block itself. The second component is a function that must be called to acknowledge Iroha block delivery. 
+The function won't make any effect if the "auto acknowledgment" mode is on. If the "auto acknowledgment" mode is off, then EVERY block must be acknowledged manually.
```
Comment (LiraLemur) [inline README.md]: have an effect. Unless the mode is on? Not very easy to understand the point. Should it be on or off?

File: README.md [hunk 5]
```diff
@@ -39,4 +40,60 @@ chain-adapter:
...
+    acknowledgment.ack();
+});
+try {
+    listener.listen();
+} catch (IllegalStateException ex) {
+    ex.printStackTrace();
+    System.exit(1);
+}
+```
+It's important to emphasize the order of calls. Calling `listen()` before defining subscribers leads to missing blocks.
```
Comment (LiraLemur) [inline README.md]: of the calls. will lead, I think

File: chain-adapter-client/src/main/java/com/d3/chainadapter/client/ReliableIrohaChainListener4J.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    public Observable<BlockSubscription> getBlockObservable() {
+        return reliableIrohaChainListener.getBlockObservable().get().map(
+                (pair) -> new BlockSubscription(pair.getFirst(), () -> pair.getSecond().invoke()));
+    }
+
+    @Override
+    public void close() throws IOException {
+        reliableIrohaChainListener.close();
+    }
+}
```
Comment (Alexey-N-Chernyshov) [inline chain-adapter-client/src/main/java/com/d3/chainadapter/client/ReliableIrohaChainListener4J.java]: 🔚

File: chain-adapter-client/src/main/kotlin/com/d3/chainadapter/client/ThreadUtil.kt [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    }.flatMap {
+        listener.listen()
+    }.fold(
+        {
+            // On listen() success
+        },
+        { ex ->
+            // On listen() failure
+        })
+}
```
Comment (Alexey-N-Chernyshov) [inline chain-adapter-client/src/main/kotlin/com/d3/chainadapter/client/ThreadUtil.kt]: 🔚 📁

**Review by Alexey-N-Chernyshov — APPROVED** [review]

File: chain-adapter-client/src/main/kotlin/com/d3/chainadapter/client/ThreadUtil.kt (hunk 7 above)
Comment (dolgopolovwork) [reply chain-adapter-client/src/main/kotlin/com/d3/chainadapter/client/ThreadUtil.kt]: Sorry. I forgot to remove it.

File: README.md (hunk 4 above)
Comment (dolgopolovwork) [reply README.md]: Unfortunately, I see no clarity issues.

**Pull request merged**